plugins/audio-s5gen2/fu-audio-s5gen2-device.c

```diff
@@ -535,24 +525,40 @@ fu_qc_s5gen2_device_write_blocks(FuQcS5gen2Device *self,
 	/* device is requesting data from the host */
 	do {
 		guint8 buf_in[FU_STRUCT_QC_DATA_REQ_SIZE] = {0};
+		gsize read_len;
 		guint32 data_sz;
 		guint32 data_offset;
 		g_autoptr(GByteArray) data_req = NULL;
 		g_autoptr(GBytes) data_out = NULL;
 
-		if (!fu_qc_s5gen2_device_msg_in(self, buf_in, sizeof(buf_in), error))
+		if (!fu_qc_s5gen2_device_msg_in(self, buf_in, sizeof(buf_in), &read_len, error))
 			return FALSE;
-		data_req = fu_struct_qc_data_req_parse(buf_in, sizeof(buf_in), 0, error);
+		data_req = fu_struct_qc_data_req_parse(buf_in, read_len, 0, error);
 		if (data_req == NULL)
 			return FALSE;
 
 		/* requested data */
 		data_sz = fu_struct_qc_data_req_get_fw_data_len(data_req);
 		data_offset = fu_struct_qc_data_req_get_fw_data_offset(data_req);
 
+		/* FIXME: abort for now */
+		if (data_sz == 0) {
+			g_autoptr(GError) error_local = NULL;
+			/* reset the update state */
+			if (!fu_qc_s5gen2_device_cmd_abort(self, &error_local))
+				g_debug("failed to abort: %s", error_local->message);
```

Comment:
should this be done in _cleanup() instead perhaps?

Reply:
hm... `_abort()` is mostly about error handling. I'd prefer to leave it in-place of potential error needed the abort.
Atm it is used in 3 places which could be handled in cleanup and 1 place which should stay as is.
Since the update could be resumed in some stages, I think the moving to cleanup() would add unneccesary complexity, imho.